flink-table/flink-sql-client/src/main/resources/keywords.properties

```diff
@@ -0,0 +1,20 @@
TWEEN;BIGINT;BINARY;BIT;BLOB;BOOLEAN;BOTH;BY;BYTES;CALL;CALLED;CARDINALITY;CASCADED;CASE;CAST;CATALOGS;CEIL;CEILING;CHAR;CHARACTER;CHARACTER_LENGTH;CHAR_LENGTH;CHECK;CLASSIFIER;CLOB;CLOSE;COALESCE;COLLATE;COLLECT;COLUMN;COMMENT;COMMIT;CONDITION;CONNECT;CONSTRAINT;CONTAINS;CONVERT;CORR;CORRESPONDING;COUNT;COVAR_POP;COVAR_SAMP;CREATE;CROSS;CUBE;CUME_DIST;CURRENT;CURRENT_CATALOG;CURRENT_DATE;CURRENT_DEFAULT_TRANSFORM_GROUP;CURRENT_PATH;CURRENT_ROLE;CURRENT_ROW;CURRENT_SCHEMA;CURRENT_TIME;CURRENT_TIMESTAMP;CURRENT_TRANSFORM_GROUP_FOR_TYPE;CURRENT_USER;CURSOR;CYCLE;DATABASES;DATE;DAY;DEALLOCATE;DEC;DECIMAL;DECLARE;DEFAULT;DEFINE;DELETE;DENSE_RANK;DEREF;DESCRIBE;DETERMINISTIC;DISALLOW;DISCONNECT;DISTINCT;DOT;DOUBLE;DROP;DYNAMIC;EACH;ELEMENT;ELSE;EMPTY;END;END-EXEC;END_FRAME;END_PARTITION;EQUALS;ESCAPE;EVERY;EXCEPT;EXEC;EXECUTE;EXISTS;EXP;EXPLAIN;EXTEND;EXTENDED;EXTERNAL;EXTRACT;FALSE;FETCH;FILTER;FIRST_VALUE;FLOAT;FLOOR;FOR;FOREIGN;FRAME_ROW;FREE;FROM;FULL;FUNCTION;FUNCTIONS;FUSION;GET;GLOBAL;GRANT;GROUP;GROUPING;GROUPS;GROUP_CONCAT;HAVING;HOLD;HOUR;IDENTITY;ILIKE;IMPORT;IN;INCLUDE;INDICATOR;INITIAL;INNER;INOUT;INSENSITIVE;INSERT;INT;INTEGER;INTERSECT;INTERSECTION;INTERVAL;INTO;IS;JOIN;JSON_ARRAY;JSON_ARRAYAGG;JSON_EXISTS;JSON_OBJECT;JSON_OBJECTAGG;JSON_QUERY;JSON_VALUE;LAG;LANGUAGE;LARGE;LAST_VALUE;LATERAL;LEAD;LEADING;LEFT;LIKE;LIKE_REGEX;LIMIT;LN;LOCAL;LOCALTIME;LOCALTIMESTAMP;LOWER;MATCH;MATCHES;MATCH_NUMBER;MATCH_RECOGNIZE;MAX;MEASURES;MEMBER;MERGE;METHOD;MIN;MINUS;MINUTE;MOD;MODIFIES;MODIFY;MODULE;MODULES;MONTH;MULTISET;NATIONAL;NATURAL;NCHAR;NCLOB;NEW;NEXT;NO;NONE;NORMALIZE;NOT;NTH_VALUE;NTILE;NULL;NULLIF;NUMERIC;OCCURRENCES_REGEX;OCTET_LENGTH;OF;OFFSET;OLD;OMIT;ON;ONE;ONLY;OPEN;OR;ORDER;OUT;OUTER;OVER;OVERLAPS;OVERLAY;PARAMETER;PARTITION;PATTERN;PER;PERCENT;PERCENTILE_CONT;PERCENTILE_DISC;PERCENT_RANK;PERIOD;PERMUTE;PIVOT;PORTION;POSITION;POSITION_REGEX;POWER;PRECEDES;PRECISION;PREPARE;PREV;PRIMARY;PROCEDURE;RANGE;RANK;RAW;READS;REAL;RECURSIVE;REF;REFERENCES;REFERENCING;REGR_AVGX;REGR_AVGY;REGR_COUNT;REGR_INTERCEPT;REGR_R2;REGR_SLOPE;REGR_SXX;REGR_SXY;REGR_SYY;RELEASE;RENAME;RESET;RESULT;RETURN;RETURNS;REVOKE;RIGHT;RLIKE;ROLLBACK;ROLLUP;ROW;ROWS;ROW_NUMBER;RUNNING;SAVEPOINT;SCALA;SCOPE;SCROLL;SEARCH;SECOND;SEEK;SELECT;SENSITIVE;SEPARATOR;SESSION_USER;SET;SHOW;SIMILAR;SKIP;SMALLINT;SOME;SPECIFIC;SPECIFICTYPE;SQL;SQLEXCEPTION;SQLSTATE;SQLWARNING;SQRT;START;STATEMENT;STATIC;STDDEV_POP;STDDEV_SAMP;STREAM;STRING;STRING_AGG;SUBMULTISET;SUBSET;SUBSTRING;SUBSTRING_REGEX;SUCCEEDS;SUM;SYMMETRIC;SYSTEM;SYSTEM_TIME;SYSTEM_USER;TABLE;TABLES;TABLESAMPLE;THEN;TIME;TIMESTAMP;TIMESTAMP_LTZ;TIMEZONE_HOUR;TIMEZONE_MINUTE;TINYINT;TO;TRAILING;TRANSLATE;TRANSLATE_REGEX;TRANSLATION;TREAT;TRIGGER;TRIM;TRIM_ARRAY;TRUE;TRUNCATE;UESCAPE;UNION;UNIQUE;UNKNOWN;UNNEST;UNPIVOT;UPDATE;UPPER;UPSERT;USE;USER;USING;VALUE;VALUES;VALUE_OF;VARBINARY;VARCHAR;VARYING;VAR_POP;VAR_SAMP;VERSIONING;VIEWS;WATERMARK;WATERMARKS;WHEN;WHENEVER;WHERE;WIDTH_BUCKET;WINDOW;WITH;WITHIN;WITHOUT;YEAR
```

Comment:
I used the code snippet below
it allows to extract some Flink specific keywords like `MATCH_RECOGNIZE` which is not a keyword in Calcite or SQL Standard e.g.
```java
public class KeywordsExtractor {
    public static void main(String[] args) {
        EnvironmentSettings es = EnvironmentSettings.newInstance().inStreamingMode().build();
        LocalStreamEnvironment lse = StreamExecutionEnvironment.createLocalEnvironment(1);
        StreamTableEnvironmentImpl ste = (StreamTableEnvironmentImpl) StreamTableEnvironmentImpl.create(lse, es);
        FlinkContext context = ((PlannerBase)ste.getPlanner()).plannerContext().getFlinkContext();

        // taken from org.apache.flink.table.planner.delegation.PlannerContext#getSqlParserConfig
        // since PlannerContext#getSqlParserConfig is private
        SqlParser.Config config = JavaScalaConversionUtil.<SqlParser.Config>toJava(
                TableConfigUtils.getCalciteConfig(context.getTableConfig()).getSqlParserConfig())
                .orElseGet(
                        // we use Java lex because back ticks are easier than double quotes in
                        // programming and cases are preserved
                        () -> {
                            SqlConformance conformance = getSqlConformance(context);
                            return SqlParser.config()
                                    .withParserFactory(FlinkSqlParserFactories.create(conformance))
                                    .withConformance(conformance)
                                    .withLex(Lex.JAVA)
                                    .withIdentifierMaxLength(256);
                        });
        SqlAbstractParserImpl parser = config.parserFactory().getParser(new SourceStringReader(""));
        System.out.println(String.join(";", parser.getMetadata().getJdbcKeywords().split(",")));
    }

    // Taken from org.apache.flink.table.planner.delegation.PlannerContext#getSqlConformance
    // since PlannerContext#getSqlConformance is private
    private static FlinkSqlConformance getSqlConformance(FlinkContext context) {
        SqlDialect sqlDialect = context.getTableConfig().getSqlDialect();
        switch (sqlDialect) {
            // Actually, in Hive dialect, we won't use Calcite parser.
            // So, we can just use Flink's default sql conformance as a placeholder
            case HIVE:
            case DEFAULT:
                return FlinkSqlConformance.DEFAULT;
            default:
                throw new TableException("Unsupported SQL dialect: " + sqlDialect);
        }
    }
}
```

Reply:
I removed keywords for `HIVE` dialect since after this PR https://github.com/apache/flink/pull/22106 keyword set for `HIVE` is the same as for `DEFAULT`

Reply:
Oops, sorry, looks like I was wrong here...
`getJdbcKeywords()` does not give all the keywords...
In fact it's a bit trickier... 
To answer a question if a whatever word is a keyword there is `org.apache.calcite.sql.parser.SqlAbstractParserImpl.MetadataImpl#isReservedWord`, e.g. 
```java
String candidate = ...;
FlinkSqlParserImpl.FACTORY
    .getParser(new StringReader(""))
    .getMetadata()
    .isReservedWord(candidate));
```

The problem is that under the hood it works like 
```java
    @Override
    public boolean isReservedWord(String token) {
      return reservedWords.contains(token);
    }
```
where `reservedWords` is a `private final Set<String> reservedWords` initialized in a constructor of `MetadataImpl` without any getter for that...

So to retrieve the list of keywords it's possible to the same code as above however in debugging mode and pick `reservedWords`'s value.


Another way with usage of `org.apache.calcite.sql.parser.SqlAbstractParserImpl.MetadataImpl#isReservedWord` has an issue that it requires lots of dependencies being shaded (calcite-core, calcite-linq4j, the whole flink-sql-parser(right now only impl part), calcite avatica, guava... )

Reply:
I think it's better we can use reflection to get these keywords. For example, we can do like this:

```
 try {
            Field reservedWords =
                    SqlAbstractParserImpl.MetadataImpl.class.getDeclaredField("reservedWords");
            reservedWords.setAccessible(true);
            return (Set<String>)
                    reservedWords.get(
                            FlinkSqlParserImpl.FACTORY
                                    .getParser(new StringReader(""))
                                    .getMetadata());
        } catch (Throwable t) {
            // ignore
            LOG.warn("Can not get reserved keywords.", t)
            return Collections.emptySet();
        }
```

Does calcite has plan to open a ticket to allow others to fetch these?